File: web/src/api/retales.ts
import { api } from '@/api/client'

export interface Retal {
  id: number
  material_categoria: string
  referencia: string
  m2_disponibles: number
  m2_original: number
  origen_numero: string
  origen_cliente: string
  fecha_ingreso: string
  estado: string
  notas: string
  precio_recuperacion: number
  precio_mercado_m2: number
}

export interface RetalIn {
  material_categoria: string
  referencia?: string
  m2_disponibles: number
  notas?: string
  precio_recuperacion?: number
  precio_mercado_m2?: number
}

export interface RetalUpdate {
  m2_disponibles?: number
  estado?: string
  notas?: string
  precio_recuperacion?: number
  precio_mercado_m2?: number
}

export async function listarRetales(): Promise<Retal[]> {
  const { data } = await api.get<Retal[]>('/api/retales')
  return data
}

export async function crearRetal(body: RetalIn): Promise<{ id: number; ok: boolean }> {
  const { data } = await api.post<{ id: number; ok: boolean }>('/api/retales', body)
  return data
}

export async function actualizarRetal(id: number, body: RetalUpdate): Promise<{ ok: boolean }> {
  const { data } = await api.put<{ ok: boolean }>(`/api/retales/${id}`, body)
  return data
}

export async function eliminarRetal(id: number): Promise<{ ok: boolean }> {
  const { data } = await api.delete<{ ok: boolean }>(`/api/retales/${id}`)
  return data
}

File: backend/routers/retales.py
from fastapi import APIRouter, Depends, Request

from backend.db.client import db_rls
from backend.middleware.auth import get_current_user
from backend.db.deps import verificar_dispositivo
from backend.models.retales import RetalIn, RetalUpdate
from backend.services import retales_service

router = APIRouter(prefix="/api/retales", tags=["retales"],
                   dependencies=[Depends(verificar_dispositivo)])


@router.get("")
def listar_retales(conn=Depends(db_rls), usuario=Depends(get_current_user)):
    return retales_service.listar_retales(conn, usuario)


@router.post("", status_code=201)
def crear_retal(body: RetalIn, request: Request, conn=Depends(db_rls),
                 usuario=Depends(get_current_user)):
    resultado = retales_service.crear_retal(
        conn, usuario, ip=request.client.host if request.client else None,
        **body.model_dump(),
    )
    return {"id": resultado["id"], "ok": True}


@router.put("/{retal_id}")
def actualizar_retal(retal_id: int, body: RetalUpdate, request: Request,
                      conn=Depends(db_rls), usuario=Depends(get_current_user)):
    cambios = body.model_dump(exclude_unset=True, exclude_none=True)
    retales_service.editar_retal(
        conn, usuario, retal_id,
        ip=request.client.host if request.client else None, **cambios,
    )
    return {"ok": True}


@router.delete("/{retal_id}")
def eliminar_retal(retal_id: int, request: Request, conn=Depends(db_rls),
                    usuario=Depends(get_current_user)):
    retales_service.eliminar_retal(
        conn, usuario, retal_id,
        ip=request.client.host if request.client else None,
    )
    return {"ok": True}
